### ISSUE TITLE
Endless loop in preprocessor due to #define A() 

### ISSUE THREAD

**Issue by albert-github** [doxygen/doxygen#7325]
In the sparse package (see: https://sparse.wiki.kernel.org/ and https://mirrors.edge.kernel.org/pub/software/devel/sparse/dist/sparse-0.6.1.tar.xz) we have a file preprocesor3.c:
```
/*
 * Each iteration of the scanning of "SCAN()" re-evaluates the recursive
 * B->A->B expansion.
 *
 * Did I already mention that the C preprocessor language
 * is a perverse thing?
 */

#define LP (

#define A() B LP )
#define B() A LP )

#define SCAN(x) x

A()                     // B ( )
SCAN( A() )             // A ( )
SCAN(SCAN( A() ))       // B ( )
SCAN(SCAN(SCAN( A() ))) // A ( )
/*
 * check-name: Preprocessor #3
 * check-description: Sparse used to get this wrong, outputting A third, not B.
 * check-command: sparse -E $file
 *
 * check-output-start

B ( )
A ( )
B ( )
A ( )
 * check-output-end
 */
```
when processing this with doxygen with the settings:
```
MACRO_EXPANSION        = YES
```
the preprocessor gets into an endless loop.
The cause of this is the fact that doxygen doesn't know the difference between `#define A()` and `#define A` and thus in the line `#define B() A LP )` tries to resolve the `A` by means of the definition of `A()`, this should not be the case.
See [What is the difference between a preprocessor macro with no arguments, and one with zero arguments](https://stackoverflow.com/questions/4251005/what-is-the-difference-between-a-preprocessor-macro-with-no-arguments-and-one-w).

Doxygen version 1.8.16 (but actually independent of the version).

The example: [example.tar.gz](https://github.com/doxygen/doxygen/files/3733864/example.tar.gz)

**Comment by doxygen** [comment]
I've added a workaround to fix the endless loop. Expansion is not yet always correctly in this corner case, but I saw no simple fix.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.17.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
